R/utilities-patterns.R

```diff
@@ -0,0 +1,100 @@
+
+#' Modify fill transparency
+#'
+#' This works much like [alpha()][scales::alpha] in that it modifies the
+#' transparency of fill colours. It differs in that `fill_alpha()` also attempts
+#' to set the transparency of `<GridPattern>` objects.
+#'
+#' @param fill A fill colour given as a `character` or `integer` vector, or as a
+#'   (list of) `<GridPattern>` object(s).
+#' @param alpha A transparency value between 0 (transparent) and 1 (opaque),
+#'   parallel to `fill`.
+#'
+#' @return A `character` vector of colours, or list of `<GridPattern>` objects.
+#' @export
+#' @keywords internal
+#'
+#' @examples
+#' # Typical colour input
+#' fill_alpha("red", 0.5)
+#'
+#' if (utils::packageVersion("grid") > "4.2") {
+#'   # Pattern input
+#'   fill_alpha(list(grid::linearGradient()), 0.5)
+#' }
+fill_alpha <- function(fill, alpha) {
+  if (!is.list(fill)) {
+    # Happy path for no patterns
+    return(alpha(fill, alpha))
+  }
+  if (is_pattern(fill) || any(vapply(fill, is_pattern, logical(1)))) {
+    check_device("patterns", action = "abort")
```

Comment:
I'm not sure this should error, honestly. The function is not called at render time so it is entirely feasible and ok to call `ggplotGrob()` while a device with no support is active, then render the grab later or save it or something

Reply:
Yes, throwing a warning is a practical middle-ground... What happens with check_device if no device is open? I'd like for it to not force one open

Reply:
You'd get the following warning:
```r
#> Warning message:
#> Unable to check the capabilities of the null device device. 
```